src/object-store/src/util.rs

```diff
@@ -44,8 +34,24 @@ pub fn normalize_dir(dir: &str) -> String {
 pub fn join_dir(parent: &str, child: &str) -> String {
     // Always adds a `/` to the output path.
     let output = format!("{parent}/{child}/");
-    // We call opendal's normalize_root which keep the last `/`.
-    opendal::raw::normalize_root(&output)
+    normalize_dir(&output)
+}
+
+/// Modified from the `opendal::raw::normalize_root`
+pub fn normalize_dir(v: &str) -> String {
+    let has_root = v.starts_with('/');
+    let mut v = v
+        .split('/')
+        .filter(|v| !v.is_empty())
+        .collect::<Vec<&str>>()
+        .join("/");
+    if has_root {
+        v.insert(0, '/');
+    }
+    if !v.ends_with('/') {
```

Comment:
We can skip this check since `v` never ends with `/`.
```suggestion
    v.push('/')
```